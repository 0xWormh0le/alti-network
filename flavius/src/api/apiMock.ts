import { application, applicationStat, peopleResponse, fileDownloadsResponse, risksResponse } from 'test/mocks'
import cloneDeep from 'lodash/cloneDeep'

const sendResponse = (response: any) => {
  return new Promise(resolve => {
    setTimeout(() => {
      resolve(cloneDeep(response))
    }, 100)
  })
}

export const APIMock = {
  getApplication: async () => await sendResponse(application),
  getApplicationStats: async () => await sendResponse(applicationStat),
  getPeople: async () => await sendResponse(peopleResponse),
  getDownloads: async () => await sendResponse(fileDownloadsResponse),
  getRisks: async () => await sendResponse(risksResponse)
}

export default { APIMock }
