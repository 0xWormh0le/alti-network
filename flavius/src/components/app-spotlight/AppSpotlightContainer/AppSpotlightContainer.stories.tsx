import React from 'react'
import { Story, Meta } from '@storybook/react/types-6-0'
import AppSpotlightContainer, { AppSpotlightContainerProps } from './AppSpotlightContainer'

export default {
  title: 'AppSpotlight/AppSpotlightContainer',
  component: AppSpotlightContainer
} as Meta

const Template: Story<AppSpotlightContainerProps> = (args) => <AppSpotlightContainer {...args} />

export const Default = Template.bind({})
Default.args = {}
