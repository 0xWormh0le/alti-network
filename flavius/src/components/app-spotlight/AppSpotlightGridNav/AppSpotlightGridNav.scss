@import 'styles/_shared.scss';

$TABITEM_WIDTH: 15em;
$TABITEM_HEIGHT: 10em;

.AppSpotlightGridTabItem {
  padding-top: $BASE_SPACING;
  width: $TABITEM_WIDTH;
  height: $TABITEM_HEIGHT;
  cursor: pointer;

  &--title {
    font-size: 14px;
    font-weight: 600;
    font-style: normal;
    line-height: 21px;
    text-align: center;
    color: $TEXT_COLOR;
  }

  &--content {
    padding-left: 34px;
    padding-top: 15px;
    @include flex(row-reverse, center, flex-start);
  }

  &--chart {
    padding-top: 15px;
  }

  &--details {
    font-family: LatoLatin;
    font-size: 10px;
    font-weight: bold;
    font-stretch: normal;
    font-style: normal;
    line-height: 1.7;
    letter-spacing: normal;
    color: $PAGINATION_SIDE_TEXT_COLOR;
    padding-right: 16px;
    padding-left: 10px;
    min-width: 90px;
    white-space: nowrap;

    ul {
      list-style: none; /* Remove default bullets */
      padding-inline-start: 10px;

      li span {
        font-weight: 400;
      }

      li::before {
        content: '\2022'; /* Add content: \2022 is the CSS Code/unicode for a bullet */
        font-weight: bold; /* If you want it to be bold */
        display: inline-block; /* Needed to add space between the bullet and the text */
        width: 1em; /* Also needed for space (tweak if needed) */
        margin-left: -1em; /* Also needed for space (tweak if needed) */
      }

      li.Severe1::before {
        color: $SEVEREITY1_BG;
      }
      li.Severe2::before {
        color: $SEVEREITY2_BG;
      }
      li.Severe3::before {
        color: $SEVEREITY3_BG;
      }
      li.Severe4::before {
        color: $SEVEREITY4_BG;
      }
      li.Severe5::before {
        color: $SEVEREITY5_BG;
      }
    }
  }
}

.SideNavNumberLoading {
  width: 32px;
  min-width: 32px;
}

.AppSpotlightTabLoading {
  display: flex;
  box-shadow: inset 0 0 2px 0 rgba(19, 70, 108, 0.25);

  &__item {
    overflow: hidden;
    min-width: $TABITEM_WIDTH;
    width: $TABITEM_WIDTH;
    height: $TABITEM_HEIGHT;
    position: relative;
  }

  &__content {
    position: absolute;
    top: 0;
    min-width: $TABITEM_WIDTH * 2;
    width: $TABITEM_WIDTH * 2;
    height: $TABITEM_HEIGHT;
    box-shadow: inset 3px 3px 2px -3px rgba(19, 70, 108, 0.25), inset 0 -3px 2px -3px rgba(19, 70, 108, 0.25);

    &.left {
      left: 0;
    }

    &.right {
      right: 0;
    }
  }
}
