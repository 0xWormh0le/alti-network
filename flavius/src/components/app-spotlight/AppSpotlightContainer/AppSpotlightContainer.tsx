import React from 'react'
import AppSpotlightAuthorizedByTable from '../AppSpotlightAuthorizedByTable'
import AppSpotlightFileDownloadsTable from '../AppSpotlightFileDownloadsTable'
import AppSpotlightAssociatedRisksTable from '../AppSpotlightAssociatedRisksTable'
import UI_STRINGS from 'util/ui-strings'
import AppSpotlightGrid from '../AppSpotlightGrid'
import { SubNavTab } from '../AppSpotlightGridNav'
import useApplicationApiClient from 'api/clients/applicationApiClient'
import { useQueryParam } from 'util/hooks'

export interface AppSpotlightContainerProps {
  wrapperType: WrapperTypeOptions
  applicationId: string
  platformId: string
}

export const navTabs: SubNavTab[] = [
  {
    label: UI_STRINGS.APPSPOTLIGHT.AUTHORIZED_BY,
    value: {
      value: 32,
      subTitleText: UI_STRINGS.APPSPOTLIGHT.OF_USERS,
      series: [13],
      serieLabels: [UI_STRINGS.APPSPOTLIGHT.INDIVIDUAL_EMPLOYEES],
      padding: 100 - 13,
      valueType: 'auth'
    },
    seriesKey: 'authorizedBy',
    title: UI_STRINGS.APPSPOTLIGHT.INDIVIDUALS_AUTHORIZED,
    unit: UI_STRINGS.APPSPOTLIGHT.USER_UNIT,
    renderDetailsTable: (appId: string, platformId: string, wrapperType: WrapperTypeOptions) => (
      <AppSpotlightAuthorizedByTable applicationId={appId} platformId={platformId} />
    )
  },
  {
    label: UI_STRINGS.APPSPOTLIGHT.FILE_DOWNLOADS,
    value: {
      value: 18,
      subTitleText: UI_STRINGS.APPSPOTLIGHT.DOWNLOADS,
      series: [18],
      serieLabels: [UI_STRINGS.APPSPOTLIGHT.ARE_SENSITIVE_FILES],
      padding: 100 - 18,
      valueType: 'download'
    },
    seriesKey: 'fileDownloads',
    title: UI_STRINGS.APPSPOTLIGHT.FILE_DOWNLOADS_BY_APP,
    unit: UI_STRINGS.APPSPOTLIGHT.DOWNLOAD_UNIT,
    renderDetailsTable: (appId: string, platformId: string, wrapperType: WrapperTypeOptions) => (
      <AppSpotlightFileDownloadsTable applicationId={appId} platformId={platformId} eventType='download' />
    )
  },
  {
    label: UI_STRINGS.APPSPOTLIGHT.ASSOCIATED_RISKS,
    value: {
      value: 15,
      subTitleText: UI_STRINGS.APPSPOTLIGHT.RISKS,
      series: [1, 2, 2, 10, 0],
      serieLabels: [5, 4, 3, 2, 1].map((magnitude) => `${UI_STRINGS.APPSPOTLIGHT.SEVERITY} ${magnitude}`),
      valueType: 'risk'
    },
    seriesKey: 'associatedRisks',
    title: UI_STRINGS.APPSPOTLIGHT.RISKS_ASSOCIATED_BY_APP,
    unit: UI_STRINGS.APPSPOTLIGHT.RISK_UNIT,
    renderDetailsTable: (appId: string, platformId: string, wrapperType: WrapperTypeOptions) => {
      return <AppSpotlightAssociatedRisksTable applicationId={appId} platformId={platformId} />
    }
  }
]

const AppSpotlightContainer: React.FC<AppSpotlightContainerProps> = (props) => {
  const { wrapperType, applicationId, platformId } = props
  const [selectedSubNavKey, setSelectedSubNavKey] = useQueryParam<AppSpotlightSubNavType>(
    'selectedSubNavKey',
    navTabs[0].seriesKey,
    { [wrapperType === 'page' ? 'page' : 'modalPage']: 1 }
  )
  const { useGetApplication, useGetApplicationStats } = useApplicationApiClient()
  const [applicationData, , applicationLoading] = useGetApplication(applicationId)
  const [applicationStatsData, , applicationStatsLoading] = useGetApplicationStats(applicationId)

  return (
    <AppSpotlightGrid
      appId={applicationId}
      platformId={platformId}
      app={applicationData}
      stats={applicationStatsData}
      loading={applicationLoading}
      statsLoading={applicationStatsLoading}
      wrapperType={wrapperType}
      navTabs={navTabs}
      selectedTab={selectedSubNavKey}
      onSubNavClick={setSelectedSubNavKey}
    />
  )
}

export default AppSpotlightContainer
