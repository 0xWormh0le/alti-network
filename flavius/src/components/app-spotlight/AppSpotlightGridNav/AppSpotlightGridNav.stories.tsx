import React from 'react'
import { Story, Meta } from '@storybook/react/types-6-0'
import AppSpotlightGridNav, { AppSpotlightGridNavProps } from './AppSpotlightGridNav'

export default {
  title: 'AppSpotlight/AppSpotlightGridNav',
  component: AppSpotlightGridNav
} as Meta

const Template: Story<AppSpotlightGridNavProps> = (args) => <AppSpotlightGridNav {...args} />

export const Risk = Template.bind({})
Risk.args = {
  subNavTab: {
    label: 'NavTab1',
    seriesKey: 'fileDownloads',
    title: 'NavTab1',
    unit: 'Unit',
    value: {
      valueType: 'risk',
      serieLabels: ['x', 'y'],
      series: [1, 2, 3, 4, 5],
      subTitleText: 'Sub title'
    } as any,
    renderDetailsTable: (appId, wrapperType) => <div />
  }
}

export const Download = Template.bind({})
Download.args = {
  subNavTab: {
    label: 'NavTab1',
    seriesKey: 'fileDownloads',
    title: 'NavTab1',
    unit: 'Unit',
    value: {
      valueType: 'download',
      serieLabels: ['x', 'y'],
      series: [1, 2, 3, 4, 5],
      subTitleText: 'Sub title'
    } as any,
    renderDetailsTable: (appId, wrapperType) => <div />
  }
}

export const Authorized = Template.bind({})
Authorized.args = {
  subNavTab: {
    label: 'NavTab1',
    seriesKey: 'fileDownloads',
    title: 'NavTab1',
    unit: 'Unit',
    value: {
      valueType: 'auth',
      serieLabels: ['x', 'y'],
      series: [1, 2, 3, 4, 5],
      subTitleText: 'Sub title'
    } as any,
    renderDetailsTable: (appId, wrapperType) => <div />
  }
}
