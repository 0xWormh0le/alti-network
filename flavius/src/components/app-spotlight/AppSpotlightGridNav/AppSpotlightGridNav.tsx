import React from 'react'
import ContentLoader from 'react-content-loader'
import DonutChart from 'components/elements/DonutChart'
import { largeNumberDisplay } from 'util/helpers'
import UI_STRINGS from 'util/ui-strings'
import './AppSpotlightGridNav.scss'

interface SubNavTabData {
  value: number
  subTitleText: string
  series: number[]
  serieLabels: string[]
  padding?: number
  valueType: 'risk' | 'download' | 'auth'
}

export interface SubNavTab {
  label: string
  seriesKey: AppSpotlightSubNavType
  value: SubNavTabData
  title: string
  unit: string
  renderDetailsTable: (appId: string, platformId: string, wrapperType: WrapperTypeOptions) => JSX.Element
}

export interface AppSpotlightGridNavProps {
  subNavTab: SubNavTab
  appStats?: any
  loading: boolean
  onSubNavClick: (key: AppSpotlightSubNavType) => void
}

const severity2 = UI_STRINGS.APPSPOTLIGHT.SEVERITY + ' 2 '
export const AppSpotlightGridTab: React.FC<AppSpotlightGridNavProps> = ({
  subNavTab,
  onSubNavClick,
  appStats,
  loading
}) => {
  const handleClick = (key: AppSpotlightSubNavType) => () => {
    onSubNavClick(key)
  }
  const totalEmails = appStats ? appStats.tileInfo.totalEmails : 0
  const totalSensitive = appStats ? appStats.tileInfo.totalSensitive : 0
  const len = appStats ? appStats.series[subNavTab.seriesKey].length : 0

  const chartData = {
    subTitleText: subNavTab.value.subTitleText,
    values: subNavTab.value.series,
    padding: -1,
    valueType: subNavTab.value.valueType,
    valueText: ''
  }

  let currentTotal: number = 0
  let currentPercentage: number = 0
  let roundedCurrentPercentage: number = 0

  switch (subNavTab.value.valueType) {
    case 'auth':
      currentTotal = appStats ? appStats.tileInfo.currentAuthorizedBy : 0
      currentPercentage = totalEmails > 0 ? currentTotal / totalEmails : 0
      roundedCurrentPercentage = Math.min(Math.round(currentPercentage * 100), 100)
      chartData.values = totalEmails > 0 ? [roundedCurrentPercentage] : [1]
      chartData.valueText = totalEmails === 0 ? '>1%' : `${roundedCurrentPercentage}%`
      chartData.padding = 100 - roundedCurrentPercentage
      break
    case 'risk':
      currentTotal = appStats ? appStats.tileInfo.currentRisks : 0
      chartData.values = currentTotal > 0 ? [currentTotal] : [1]
      chartData.valueText = largeNumberDisplay(currentTotal)
      break
    case 'download':
      currentTotal =
        len > 0 ? appStats.series[subNavTab.seriesKey].reduce((total: number, current: number) => total + current) : 0
      chartData.valueText = largeNumberDisplay(currentTotal)
      currentPercentage = totalSensitive > 0 ? totalSensitive / currentTotal : 0
      roundedCurrentPercentage = Math.min(Math.round(currentPercentage * 100), 100)
      chartData.values = totalSensitive > 0 ? [roundedCurrentPercentage] : [1]
      chartData.padding = 100 - roundedCurrentPercentage
      break
    default:
      const _exhasutiveCheck: never = subNavTab.value.valueType
      throw new Error(`Attempted to render an unexpected App Spotlight Nav Tab type: ${_exhasutiveCheck}`)
  }

  const detailElement = () => {
    if (loading) {
      return <TabContentLoading />
    }
    switch (subNavTab.value.valueType) {
      case 'auth':
        return UI_STRINGS.APPSPOTLIGHT.OF_TOTAL_USERS(largeNumberDisplay(totalEmails))
      case 'risk':
        return (
          <ul>
            <li className='Severe2'>
              {severity2}
              <span>
                {currentTotal} {currentTotal === 1 ? 'risk' : 'risks'}
              </span>
            </li>
            {/* {subNavTab.value.serieLabels.map((label, index) => (
              <li key={index} className={cx('Severe' + (5 - index))}>
                {label} {subNavTab.value.series[index]} {subNavTab.value.series[index] > 1 ? 'risks' : 'risk'}
              </li>
            ))} */}
          </ul>
        )
      default:
        return UI_STRINGS.APPSPOTLIGHT.TOTAL_SENSITIVE_FILES(totalSensitive)
    }
  }

  return (
    <div key={subNavTab.seriesKey} className='AppSpotlightGridTabItem' onClick={handleClick(subNavTab.seriesKey)}>
      <div className='AppSpotlightGridTabItem--title'>{subNavTab.label}</div>
      <div className={`AppSpotlightGridTabItem--content`}>
        <div className='AppSpotlightGridTabItem--details'>{detailElement()}</div>
        <div className='AppSpotlightGridTabItem--chart'>
          <DonutChart data={chartData} loading={loading} />
        </div>
      </div>
    </div>
  )
}

const TabContentLoading = () => (
  <div className='TabContentLoading'>
    <ContentLoader
      backgroundColor='#F0F0F0'
      foregroundColor='#F7F7F7'
      height={29}
      width={79}
      uniqueKey='TabContentLoading'>
      <rect x={0} y={0} width={79} height={29} rx={4} ry={4} />
    </ContentLoader>
  </div>
)

export default AppSpotlightGridTab
