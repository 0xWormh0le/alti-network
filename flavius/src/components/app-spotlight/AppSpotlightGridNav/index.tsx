import { SubNavTab as T } from './AppSpotlightGridNav'
export type SubNavTab = T

export { default } from './AppSpotlightGridNav'
