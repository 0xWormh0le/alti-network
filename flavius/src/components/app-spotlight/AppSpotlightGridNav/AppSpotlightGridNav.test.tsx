import React from 'react'
import { renderWithRouter } from 'test/support/helpers'
import noop from 'lodash/noop'
import AppSpotlightGridNav from './AppSpotlightGridNav'
import { navTabs } from '../AppSpotlightContainer/AppSpotlightContainer'
import apiMock from 'api/apiMock'

describe('AppSpotlightGridNav', () => {
  it('renders correctly', async () => {
    const appStats = await apiMock.APIMock.getApplicationStats()

    const { container } = renderWithRouter(
      <AppSpotlightGridNav appStats={appStats} subNavTab={navTabs[0]} loading={false} onSubNavClick={noop} />
    )

    expect(container).toMatchSnapshot()
  })
})
