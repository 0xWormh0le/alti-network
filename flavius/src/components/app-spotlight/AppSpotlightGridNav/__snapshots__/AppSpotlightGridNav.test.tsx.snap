// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppSpotlightGridNav renders correctly 1`] = `
<div>
  <div
    class="AppSpotlightGridTabItem"
  >
    <div
      class="AppSpotlightGridTabItem--title"
    >
      Authorized By
    </div>
    <div
      class="AppSpotlightGridTabItem--content"
    >
      <div
        class="AppSpotlightGridTabItem--details"
      >
        of 100 total users
      </div>
      <div
        class="AppSpotlightGridTabItem--chart"
      >
        <div
          class="DonutChart DonutChart__auth"
        >
          <div
            class="ct-chart "
          />
          <div
            class="DonutChart__body"
          >
            <div
              class="DonutChart__body--title"
            >
              90%
            </div>
            <div
              class="DonutChart__body--subtitle"
            >
              of Users
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
